### DISCUSSION TITLE
Nucleo-L4R5 is kept in reset after downloading sketch

### DISCUSSION THREAD

**Discussion by PerennialNovice** [stm32duino/Arduino_Core_STM32#1517] [General]
Hi,

I am experimenting with Nucleo-L4R5 and try to run the [CheckVariant example sketch](https://github.com/stm32duino/STM32Examples/tree/main/examples/NonReg/CheckVariant) on it.
This sometime succeedes, but most of the time the µC gets stuck in Reset. The red LED is on. When trying to program it with STM32CubeProgramer I get the error "The core is kept under Reset!"
Doing a full chip erase and several resets is able to bring the µC back to normal operation but this just keeps happening very often.
I also tried downloading a simple "blink" example, same result.
When connecting the stuck board to STM32CubeProgrammer, fault analysis is executed and I get the message that DIV_0_TRP bit was enabled.
I suspect this is an error with stm32duino, not with the µC itself, thats why I am writing here.

**Comment by PerennialNovice** [comment]
Same behaviour on Windows and Linux (Xubuntu in VM)

**Comment by fpistm** [comment]
Hi @Pestimist 
in fact this is not linked to the core.
Issue is under investigation internally as it has been reproduced using application generated thanks STM32CubeIDE.
It has been found with SWAN R5 board, see https://github.com/stm32duino/Arduino_Core_STM32/pull/1511
As a workaround you can simply use the mass storage upload method.

**Comment by PerennialNovice** [comment]
Thank you @fpistm for this information
mass storage upload may help here but is not available for the generic variants (I'm still trying to create the L4P5 variant). Is there a possibility to have it available for generic variants?

**Comment by fpistm** [comment]
Unfortunately not it is too specific to be added to generic upload method.

**Comment by PerennialNovice** [comment]
too bad
can I make use of the fix you proposed? How could I do that?

**Comment by fpistm** [comment]
Well you can locally change the GenL4 upload line to add: -ob nBOOT0=0 -ob nBOOT0=1
https://github.com/stm32duino/Arduino_Core_STM32/blob/6479b8a0936852c052c16f79606cd1192fc9dcc1/boards.txt#L5089
I did not add it as issue seems linked to STM32L4+ and do not want impact all the STM32L4.

**Comment by PerennialNovice** [comment]
Thank you very much

**Comment by PerennialNovice** [comment]
After successfully downloading the CheckVariant sketch via SWD the µC started to behave very erratically.

First it was unreacheable, found out that Flash protection registers were changed.
Managed to get them back to normal and program the chip anew.
But the registers keep getting messed up on nearly every programming.
And the chip keeps resetting, output on terminal is as follows:
```
#####
NUM_DIGITAL_PINS#####
NUM_DIGITAL_PINS#####
NUM_DIGITAL_PINS#####
NUM_DIGITAL_PINS#####
NUM_DIGITAL_PINS#####
NUM_DIGITAL_PINS#####
NUM_DIGITAL_PINS#####
NUM_DIGITAL_PINS#####
NUM_DIGITAL_PINS#####
NUM_DIGITAL_PINS#####
NUM_DIGITAL_PINS#####
```

This happens regardless how I program the chip (Arduino IDE, Sloeber, STM32programmer, mass storage, SWD)
Even writing registers and resetting the core within STM32Cube Programmer yields the wrong entries in the registers.
seems like the chip itself is broken?

**Comment by fpistm** [comment]
Yes very strange.
Anyway, what do you want achieve? Nucleo L4R5 is already available in the core?

**Comment by PerennialNovice** [comment]
I wanted to add the L4P5 variant, did all the steps mentioned [here](https://github.com/stm32duino/wiki/wiki/Add-a-new-variant-%28board%29) but had no serial output after programming the CheckVariant sketch.
So I tried the sketch wit the L4R5 variant just to make sure all the rest of my setup is working fine.
That is how I ran into all these issues.

**Comment by fpistm** [comment]
So you tried to add the generic L4P5.  If Serial does not work probably that generic  SystemCoreConfig is not correct or ldscript as this is what is missing (and off course boards.txt entry addition)

**Comment by PerennialNovice** [comment]
Yes, but now I also want to get the Nucleo-L4R5 back working, if that is possible.
And I really would like to know what is wrong with it and how I can possibly have caused that. 
I don't want to brake another board...

**Comment by fpistm** [comment]
Well, hard to tell what is wrong. You can try to use the STM32CubeProgrammer and check options bytes.
Also update the STLink Firmware of your board if needed.

**Comment by PerennialNovice** [comment]
ST-Link firmware update did not change anything.
To be honest I don't really know much about the option bytes, and there are many :)
Anything special you consider worth looking at?
Thanks for your support!

**Comment by fpistm** [comment]
You can share a screenshot of your OB then I will compare to mine.

**Comment by PerennialNovice** [comment]
![image](https://user-images.githubusercontent.com/85557472/137687287-0d336a2c-8ab3-48a0-b9f0-a3a54e368218.png)
![image](https://user-images.githubusercontent.com/85557472/137687326-25db1706-8c3a-4b4a-accb-a1864e4f8e30.png)
![image](https://user-images.githubusercontent.com/85557472/137687354-f031bedf-9be6-4dac-9789-15fbf30da943.png)

**Comment by PerennialNovice** [comment]
This seems to be when all works fine, trying to recreate the "not working" state

**Comment by PerennialNovice** [comment]
side question:
is there a difference between Windows and Linux versions of Arduino IDE and STM32duino considering this issue?
I'm asking because build on Linux is way faster even when running out of a VM...
would save me some time

**Comment by PerennialNovice** [comment]
OB are the same when the MCU doesn't start and the red LED on the Nucleo board is lit

**Comment by fpistm** [comment]
I've the same. Issue come from Arduino IDE on Windows. Core is only source file.
On windows you can try to disable your antivirus.

> This seems to be when all works fine, trying to recreate the "not working" state

I thought your board does not work anymore ?

**Comment by PerennialNovice** [comment]
it is complicated...
I don't know why, but sometimes i can download a sketch (or bin via CubeProgrammer) and the code gets executed. Sometimes the red LED is on and the board seems to do nothing
still trying to find out what causes these different states, no luck so far :/

**Comment by fpistm** [comment]
I would checked nBoot0 and nSWBoot0 by default.
About the issue it is similar to the one I've mentioned before and I don't have more clue for you as it is not related to the core and under investigation internally by our experts but it can take long time to have their feedback.

**Reply by fpistm** [reply]
I moved this thread as a discussion as it is not linked to the core (as I said I've reproduced with a simple blink project with STM32CubeIDE).

**Comment by PerennialNovice** [comment]
...now I finally understand that with "core" you aer NOT talking about the MCU core, but the stm32duino-core...

With nBOOT0 and nSWBOOT0 checked **and** BFB2 _unchecked_ it seems to work reliably (I compared OB with the ones from Nucleo-L4P5 I also have)

**Comment by fpistm** [comment]
Right in Arduino world core is defined like this: https://www.arduino.cc/en/guide/cores
When I talk about the hardware component, I talk about MCU.

**Comment by CHAMSTM** [comment]
Hello,

I'm  not sure if my answer is still useful now, but:
Here the issue is that when you write a binary in the flash and try to execute it, it doesn't execute.
This is normal due to PEMPTY that is cleared only by power off/on reset.

The flash is considered "empty" as long as first word @0x0800 0000 contains 0xFFFFFFFF.
The evaluation of this word is done at device startup (after reset) and PEMPTY is set if condition is true.
If PEMPTY is set, then the boot is forced (by hardware) to the system Bootloader (0x1FFFxxxx)
PEMPTY can be cleared only after a power off/on reset (not after a system reset).

So if you take an empty device, write your binary then reset (using any tool), it will not execute your loaded code. It will execute System Bootloader till next power off/on reset.

You can get rid of this behavior by forcing boot on Flash using Option bytes nBootSW=0, nBoot0=1, nBoot1=1.

I hope it helps.

**Reply by PerennialNovice** [reply]
I haven't had time to play aournd with the Nucleos lately but every bit of information is always welcome :)
Especially if I learn something new.
thank you very much for sharing your insights :)

**Reply by fpistm** [reply]
> Hello,
> 
> I'm not sure if my answer is still useful now, but: Here the issue is that when you write a binary in the flash and try to execute it, it doesn't execute. This is normal due to PEMPTY that is cleared only by power off/on reset.
> 
> The flash is considered "empty" as long as first word @0x0800 0000 contains 0xFFFFFFFF. The evaluation of this word is done at device startup (after reset) and PEMPTY is set if condition is true. If PEMPTY is set, then the boot is forced (by hardware) to the system Bootloader (0x1FFFxxxx) PEMPTY can be cleared only after a power off/on reset (not after a system reset).
> 
> So if you take an empty device, write your binary then reset (using any tool), it will not execute your loaded code. It will execute System Bootloader till next power off/on reset.
> 
> You can get rid of this behavior by forcing boot on Flash using Option bytes nBootSW=0, nBoot0=1, nBoot1=1.
> 
> I hope it helps.

Thanks @CHAMSTM for your feedback. Anyway it only explain for the first flash.
Issue here is that for the same Nucleo L4R5 after a first correct programming it works or not. Same issue ss stated in the internal BZ.
